---
title: "Analysis of Psoriasis scRNA-Seq data from Bennet"
author: "Maria Schmidt"
affiliation: "IZBI, Interdisciplinary Centre for Bioinformatics, Leipzig University, Germany"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    number_sections: false
bibliography: "my_collection.bib"
link-citations: true
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE)

library(Seurat)
library(tidyverse)
library(scales)
library(patchwork)
library(Matrix)
# library(clustree)
library(ggplot2)
library(pals)
library(distill)

source("/scratch/maria/Functions/liana_dotplots.r")

load("/scratch/maria/Benett Frost/BF-01-02-03-all-Analysis/BF-01-02-03-all-Analysis_seuratObj_integrated_percent.mt<15.RData")
load("/scratch/maria/Benett Frost/BF-01-02-03-all-Analysis/BF-01-02-03-all-Analysis_seuratObj_my.palettes.RData")
load("/scratch/maria/Benett Frost/BF-01-02-03-all-Analysis/BF-01-02-03-all-Analysis_BennetsSkinMarker.RData")
load("/scratch/maria/Benett Frost/BF-01-02-03-all-Analysis/BF-01-02-03-all-Analysis_seuratObj_integrated_predicted.id.Cell_type.markers.RData")
load("/scratch/maria/Benett Frost/BF-01-02-03-all-Analysis/Reynolds21_skin.marker.RData")
load("/scratch/maria/Benett Frost/BF-01-02-03-all-Analysis/BF-01-02-03-all-Analysis_liana.obj_reynolds21.cell.type.RData")
load("BF-01-02-03-all-Analysis_seuratObj_integrated_TXLNA_cells_markers.RData")
load("BF-01-02-03-all-Analysis_seuratObj_integrated_IFNL1_cells_markers.RData")
load("/scratch/maria/Benett Frost/interleukins.RData")
interleukins <- unique(interleukins, rownames(obj_integr))

DefaultAssay(obj_integr) <- "integrated"

cluster.colors <- polychrome(length(levels(obj_integr@meta.data$seurat_clusters)))
names(cluster.colors) <- levels(obj_integr@meta.data$seurat_clusters)
celltype.colors <- my_palettes[["predicted.id.Cell_type"]]
names(celltype.colors) <- unique(obj_integr$predicted.id.Cell_type_short)
Idents(obj_integr) <- "orig.ident"
celltype.cluster.id <- "predicted.id.Cell_type_short"
celltype_cluster_data <- data.frame(
  table(obj_integr@meta.data[, "predicted.id.Cell_type_short"]),
  "color" = celltype.colors
)
seurat.cluster.data <- data.frame(
  table(obj_integr@meta.data[, "seurat_cluster"]),
  "color" = scales::hue_pal()(length(table(obj_integr@meta.data[, "seurat_cluster"])))
)
```

# Preprocessing, data integration and dimension reduction

## Quality check

We checked quality of all 3 Samples and removed cells with high mitochondrial expression (\> 15%) and with to many counts (\> median + 3 Standard Deviations), so that we filtered from 3772 to `r length(obj_integr$seurat_clusters)` cells. Filtering out low count cells was done by cellranger.

```{r, echo=FALSE}
VlnPlot(obj_integr, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"))
FeatureScatter(obj_integr, feature1 = "nCount_RNA", feature2 = "percent.mt")
ggplot(obj_integr@meta.data) +
    geom_point(aes(x = nCount_RNA, y = nFeature_RNA, colour = percent.mt)) +
    theme_minimal()
```

## Data integration and dimension reduction

We integrated the 3 datasets into one, reduced Dimension with PCA and visualized results with UMAP.

```{r, echo=FALSE}
DimPlot(obj_integr, reduction = "umap", label = FALSE, group.by = "orig.ident")
```

## Cell-cycle scoring

We assigned each cell a cell-cycle score, based on its expression of G2/M and S phase markers and predicted classification of each cell in either G2M, S or G1 phase. Cell cycle scoring showed moderate cell cycle effect, mostly of cycling Keratinocytes.

```{r, echo=FALSE,  fig.width = 6, fig.height = 5}
DimPlot(obj_integr, reduction = "umap", label = FALSE, group.by = "Phase")
```

## Unsupervised louvain clustering

To cluster cell in an unsupervised fashion, we applied the Louvain-algorithm implemented in the Seurat package.

```{r, echo=FALSE,  fig.width = 10, fig.height = 5}
p1 <- DimPlot(obj_integr, reduction = "umap", label = TRUE, repel = TRUE, group.by = "seurat_clusters") + theme(legend.position = "none")
p2 <- ggplot(seurat.cluster.data, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  scale_fill_manual(values = seurat.cluster.data[, "color"]) +
  coord_flip() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
  ) +
  xlab("Cluster") +
  ylab("Frequency")

p1 + p2
```

## Celltype classification using a reference dataset

We projected celltypes from a healthy scRNA-Seq dataset (Reynolds et al., 2021) onto our dataset.

```{r, echo=FALSE,  fig.width = 10, fig.height = 5}

Idents(obj_integr) <- "predicted.id.Cell_type_short"

p1 <- DimPlot(obj_integr, reduction = "umap", group.by = "predicted.id.Cell_type_short", cols = celltype.colors) + theme(legend.position = "none", legend.text = element_text(size = 7)) + ggtitle("")

p2 <- ggplot(celltype_cluster_data, aes(x = Var1, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  scale_fill_manual(values = celltype.colors) +
  coord_flip() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
  ) +
  xlab("Celltype") +
  ylab("Frequency")

p1 + p2
```

To verify this further, we examined some canonical cell type markers for the specific skin and immune cell populations.

```{r, echo=FALSE, fig.align = "center", fig.height = 10, fig.width = 15 }
DotPlot(obj_integr, features = skin_marker, group.by = "predicted.id.Cell_type_short", dot.scale = 6) +
  scale_color_gradient2(high = "red", low = "blue") +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
  )
```

## Celltypes per dataset {.tabset}

### BF-LE-01-KT-PSO_all

```{r, echo=FALSE, fig.width = 10, fig.height = 5}
Idents(obj_integr) <- "orig.ident"

plotting_subset <- function(my_ident = my_ident, seurat_object = seurat_object, celltype_cluster_data = celltype_cluster_data){
  
  Idents(seurat_object) <- "orig.ident"
  my_subset <- subset(seurat_object, idents=my_ident)
  Idents(my_subset) <- "predicted.id.Cell_type_short"
  celltype_cluster_data_subset <- data.frame(
    table(my_subset@meta.data[, "predicted.id.Cell_type_short"]),
    "color" = celltype.colors[names(table(my_subset@meta.data[, "predicted.id.Cell_type_short"]))]
  )
  
  p1 <- DimPlot(my_subset, reduction = "umap", group.by = "predicted.id.Cell_type_short", cols = celltype.colors) + theme(legend.position = "none", legend.text = element_text(size = 7)) + ggtitle("")
  
  p2 <- ggplot(celltype_cluster_data_subset, aes(x = Var1, y = Freq, fill = Var1)) +
    geom_bar(stat = "identity") +
    theme_minimal() +
    scale_fill_manual(values = celltype_cluster_data_subset$color) +
    coord_flip() +
    theme(
      legend.position = "none",
      axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
    ) +
    xlab("Celltype") +
    ylab("Frequency")
  
  p1 + p2
}
plotting_subset(my_ident = unique(Idents(obj_integr))[1], seurat_object = obj_integr, celltype_cluster_data = celltype_cluster_data)
```

### BF-LE-02-PG-PSO_all


```{r, echo=FALSE, fig.width = 10, fig.height = 5}

plotting_subset(my_ident = unique(Idents(obj_integr))[2], seurat_object = obj_integr, celltype_cluster_data = celltype_cluster_data)

```

### BF_LE_03_VC_03_all

```{r, echo=FALSE, fig.width = 10, fig.height = 5}

plotting_subset(my_ident = unique(Idents(obj_integr))[3], seurat_object = obj_integr, celltype_cluster_data = celltype_cluster_data)

```

# Differentially expressed features

We calculated positive markers via differential expression, which further verify and define our predicted clusters .

```{r, echo=FALSE, fig.align = "center", fig.height = 20, fig.width = 20 }
top5_predicted.id.Cell_type <- predicted.id.Cell_type.marker %>%
  group_by(cluster2) %>%
  top_n(n = 5, wt = avg_log2FC)
DoHeatmap(obj_integr, features = top5_predicted.id.Cell_type$gene, group.by = "predicted.id.Cell_type_short", group.bar.height = 0.03, group.colors = celltype.colors, hjust = 0) +
  xlab("Celltype") +
  ylab("Top 5 Marker per Cluster") + theme(legend.position = "none") +
  theme(
    axis.text.y = element_text(size = 15),
    axis.title = element_text(size = 30),
    axis.text.x = element_text(size = 25)
  )
```

# Interleukines (ILs)

## Which ILs where expressed and in which Cluster?

### All differentilly expressed ILs across all celltypes

```{r, echo=FALSE, fig.align = "center", fig.height = 9, fig.width = 15 }
interleukins.diffexpr <- unique(predicted.id.Cell_type.marker$gene[which(predicted.id.Cell_type.marker$gene %in% interleukins)])

DotPlot(obj_integr, features = interleukins.diffexpr, group.by = "predicted.id.Cell_type_short", dot.scale = 5) +
  scale_color_gradient2(high = "red", low = "blue") +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
  )
```

### All expressed ILs across all celltypes

```{r, echo=FALSE, fig.align = "center", fig.height = 8, fig.width = 20 }

interleukins <- rownames(obj_integr)[which(rownames(obj_integr) %in% interleukins)]

DotPlot(obj_integr, features = interleukins, group.by = "predicted.id.Cell_type_short", dot.scale = 6) +
  scale_color_gradient2(high = "red", low = "blue") +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 14),
    axis.text.y = element_text(size = 16)
  )
```

## Which cells express *IL-14 (TXLNA)* and *IL-29 (IFNL1)*?

```{r, echo=FALSE, fig.align = "center", fig.height = 5, fig.width = 10 }
genes.bennet <- c("TXLNA", "IFNL1")
Idents(obj_integr) <- "predicted.id.Cell_type_short"

                        FeaturePlot(
                            obj_integr, 
                            features = genes.bennet, 
                            order = TRUE, 
                            label = TRUE, 
                            repel = TRUE, 
                            min.cutoff = "q1", 
                            max.cutoff = "q99"
                        ) 
```

```{r, echo=FALSE, fig.align = "center", fig.height = 5, fig.width = 8 }
DotPlot(obj_integr, features = genes.bennet, group.by = "predicted.id.Cell_type_short", dot.scale = 4) +
  scale_color_gradient2(high = "red", low = "blue") +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
  )

```

### Marker of *TXLNA*-UP and *IFNL1*-UP cells {.tabset}

#### *TXLNA*-UP cells

```{r, echo=FALSE, fig.align = "center", fig.height = 12, fig.width = 10}

# TXLNA_UP <- mean(GetAssayData(object = obj_integr, slot = "data")["TXLNA",])
# sd_TXLNA_UP <- sd(GetAssayData(object = obj_integr, slot = "data")["TXLNA",])
# obj_TXLNA <- subset(x = obj_integr, subset = TXLNA > (TXLNA_UP+sd_TXLNA_UP))
# obj_integr$TXLNA_expr <- "-"
# obj_integr$TXLNA_expr[ colnames(obj_TXLNA)] <- "up"
# Idents(obj_integr) <- "TXLNA_expr"
# TXLNA_cells_markers <- FindMarkers(object = obj_integr, ident.1 = "up")
# save(TXLNA_cells_markers, file = "BF-01-02-03-all-Analysis_seuratObj_integrated_TXLNA_cells_markers.RData")
head(TXLNA_cells_markers, 10)
TXLNA_coexpr <- head(rownames(TXLNA_cells_markers), 7)[-1]

FeaturePlot(
    obj_integr,
    features = TXLNA_coexpr,
    order = TRUE,
    label = TRUE,
    repel = TRUE,
    min.cutoff = "q1",
    max.cutoff = "q99"
)
```

#### *IFNL1*-UP cells

```{r, echo=FALSE, fig.align = "center", fig.height = 12, fig.width = 10 }

# IFNL1_UP <- mean(GetAssayData(object = obj_integr, slot = "data")["IFNL1",])
# sd_IFNL1_UP <- sd(GetAssayData(object = obj_integr, slot = "data")["IFNL1",])
# obj_IFNL1 <- subset(x = obj_integr, subset = IFNL1 > (IFNL1_UP+sd_IFNL1_UP))
# obj_integr$IFNL1_expr <- "-"
# obj_integr$IFNL1_expr[ colnames(obj_IFNL1)] <- "up"
# Idents(obj_integr) <- "IFNL1_expr"
# IFNL1_cells_markers <- FindMarkers(object = obj_integr, ident.1 = "up")
# save(IFNL1_cells_markers, file = "BF-01-02-03-all-Analysis_seuratObj_integrated_IFNL1_cells_markers.RData")
head(IFNL1_cells_markers, 10)
IFNL1_coexpr <- head(rownames(IFNL1_cells_markers), 7)[-1]

FeaturePlot(
    obj_integr,
    features = IFNL1_coexpr,
    order = TRUE,
    label = TRUE,
    repel = TRUE,
    min.cutoff = "q1",
    max.cutoff = "q99"
)

```

## Expression of further selected ILs/Features

```{r, echo=FALSE, fig.align = "center",  fig.height = 24, fig.width = 10 }
genes.bennet <- c("IL17A", "IL17F", "TNF", "TNFRSF1A", "IL22", "IL12A", "IL12B", "IL23R", "IL23A", "NOS1", "NOS2", "IL13", "IL13RA1", "IL17RA")
Idents(obj_integr) <- "predicted.id.Cell_type_short"
FeaturePlot(
  obj_integr, 
  features = genes.bennet,
  ncol = 2,
  order = TRUE, 
  label = TRUE, 
  repel = TRUE, 
  min.cutoff = "q1",
  max.cutoff = "q99"
  )
```

```{r, echo=FALSE, fig.align = "center", fig.height = 7, fig.width = 7}
DotPlot(obj_integr, features = genes.bennet, group.by = "predicted.id.Cell_type_short", dot.scale = 4) +
  scale_color_gradient2(high = "red", low = "blue") +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
  )

```

## Expression of all ILs in Tc, Th and Treg cells

```{r, echo=FALSE, fig.align = "center", fig.height = 10, fig.width = 7}
cells <- colnames(obj_integr)[which(obj_integr$predicted.id.Cell_type_short %in% c("Th", "Tc", "Treg"))]
DoHeatmap(obj_integr, features = interleukins, group.by = "predicted.id.Cell_type_short", cells = cells, group.colors = celltype.colors)
```
## Annotation of Th1 and Th2 subpopulations {.tabset}

### Th1 marker

```{r, echo=FALSE, fig.align = "center", fig.height = 10, fig.width = 7}
th_cells <- colnames(obj_integr)[which(obj_integr$predicted.id.Cell_type_short %in% c("Th"))]
th1_marker <- c("STAT1", "STAT2", "IFNG", "IL2", "TNF", "TNF")
tcells_obj <- subset(obj_integr, cells = th_cells)
FeaturePlot(
  tcells_obj,
  features = th1_marker,
  ncol = 2,
  order = TRUE, 
  min.cutoff = "q1",
  max.cutoff = "q99"
  )
# my_subset <- subset(obj_integr, cells = cells) %>%
#                    RunPCA(features = c(th1_marker, th2_marker), verbose = FALSE) %>%
#                    RunUMAP(features = c(th1_marker, th2_marker)) %>%
#                    FindNeighbors() %>%
#                   FindClusters()
# DimPlot(my_subset)                     
# FeaturePlot(my_subset, features = th1_marker)                     
# FeaturePlot(my_subset, features = th2_marker[9:length(th2_marker)])                     
```
### Th2 marker

```{r, echo=FALSE, fig.align = "center", fig.height = 21, fig.width = 10}
th2_marker <- c("IL4", "IL5", "IL9", "IL10", "IL13", "IL21", "GATA3", "IRF4", "STAT5A", "STAT5B","STAT6")
FeaturePlot(
  tcells_obj, 
  features = th2_marker,
  ncol = 2,
  order = TRUE, 
  min.cutoff = "q1",
  max.cutoff = "q99"
  )
```

## UMAP colored by expression of ILs differentially expressed in T-cells

```{r, echo=FALSE, fig.align = "center", fig.height = 15, fig.width = 10 }

interleukins.tcells <- predicted.id.Cell_type.marker[grep("^IL", predicted.id.Cell_type.marker$gene), ]
interleukins.tcells <- unique(interleukins.tcells[which(interleukins.tcells$cluster2 %in% c("Th", "Tc", "Treg")), "gene"])
Idents(obj_integr) <- "predicted.id.Cell_type_short"
FeaturePlot(obj_integr, features = interleukins.tcells, order = TRUE, ncol = 2, label = TRUE, repel = TRUE, min.cutoff = "q1", max.cutoff = "q99")
```

# Receptor-Ligand Analysis (Ofosu-Bamfo et al., 2019)

## Lymphocytes \<-\> Lymphocytes

```{r, echo=FALSE, fig.align = "center", fig.height = 15, fig.width = 20 }
lymphocytes <- c("Tc", "Th", "Treg", "Plasma cell")
tcells <- c("Tc", "Th", "Treg")

liana_subset <- liana_test  %>% 
  filter(aggregate_rank < 0.05 )
  if(nrow(liana_subset) > 0){
  		my_liana_dotplot2(liana_subset, source_groups = lymphocytes,
  					target_groups = lymphocytes)  %>%
  		print()
  }

```

## Keratinocytes -\> Lymphocytes

```{r, echo=FALSE, fig.align = "center", fig.height = 9, fig.width = 20 }
keratinocytes <- c("KC premit", "KC_postmit", "KC_mit")

liana_subset <- liana_test  %>% 
  filter(aggregate_rank < 0.05 )
  if(nrow(liana_subset) > 0){
  		my_liana_dotplot2(liana_subset, source_groups = keratinocytes,
  					target_groups = lymphocytes)  %>%
  		print()
  }

```

## Lymphocytes -\> Keratinocytes

```{r, echo=FALSE, fig.align = "center", fig.height = 10, fig.width = 16 }
keratinocytes <- c("KC premit", "KC_postmit", "KC_mit")

liana_subset <- liana_test  %>% 
  filter(aggregate_rank < 0.05 )
  if(nrow(liana_subset) > 0){
  		my_liana_dotplot2(liana_subset, source_groups = lymphocytes,
  					target_groups = keratinocytes)  %>%
  		print()
  }

```


## T cells -\> Dendritic cells

```{r, echo=FALSE, fig.align = "center", fig.height = 10, fig.width = 16 }
dendritic <- grep("DC", unique(obj_integr$predicted.id.Cell_type_short), value = TRUE)
tcells <- c("Tc", "Th", "Treg")

liana_subset <- liana_test  %>% 
  filter(aggregate_rank < 0.05 )
  if(nrow(liana_subset) > 0){
  		my_liana_dotplot2(liana_subset, source_groups = tcells,
  					target_groups = dendritic)  %>%
  		print()
  }

```
## T cells -\> Langerhans cells

```{r, echo=FALSE, fig.align = "center", fig.height = 10, fig.width = 14 }
langerhans <- grep("LC", unique(obj_integr$predicted.id.Cell_type_short), value = TRUE)
tcells <- c("Tc", "Th", "Treg")

liana_subset <- liana_test  %>% 
  filter(aggregate_rank < 0.05 )
  if(nrow(liana_subset) > 0){
  		my_liana_dotplot2(liana_subset, source_groups = tcells,
  					target_groups = langerhans)  %>%
  		print()
  }

```