---
title: "Analysis of Psoriasis scRNA-Seq data from Bennet"
author: "Maria Schmidt"
affiliation: "IZBI, Interdisciplinary Centre for Bioinformatics, Leipzig University, Germany"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    number_sections: false
link-citations: true

---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo = FALSE)
```

```{r}
pacman::p_load(Seurat, tidyverse, here, ggplot2, pals, flextable, xlsx)
source(here("src","seurat-functions.r"))
source(here("src","make-nicknames-for-celltypes.r"))
output_data_path <- "output"
input_data_path <- "data"

filename <- here(output_data_path, "Reynolds2021_Data-all_seuratObj.RData")
load(filename)
seurat_object$predicted.id.Cell_type_nicknames <- make_nicknames(seurat_object$Cell_type)
obj_reynolds <- seurat_object
Idents(obj_reynolds) <- "predicted.id.Cell_type_nicknames"
obj_reynolds <- subset(obj_reynolds, downsample = 100)

filename <- here(output_data_path, "integrated-seurat-obj.RData")
load(filename)
filename <- here(output_data_path, "color-list.RData")
load(filename)  
celltype_colors <- color_list[["predicted.id.Cell_type_nicknames"]]
seurat_colors <- scales::hue_pal()(length(table(obj_integr@meta.data[, "seurat_clusters"])))

filename <- here(output_data_path, "BF_LE_06_KS_LE_all_seurat-obj.RData")
load(filename)
seurat_object$predicted.id.Cell_type_nicknames <- make_nicknames(seurat_object$predicted.id.Cell_type)
obj_06 <- seurat_object

filename <- here(output_data_path, "BF_LE_08_GD_pre_h_seurat-obj.RData")
load(filename)
seurat_object$predicted.id.Cell_type_nicknames <- make_nicknames(seurat_object$predicted.id.Cell_type)
obj_08 <- seurat_object

filename <- here(output_data_path, "interleukins.RData")
load(filename)

load(here(output_data_path, "integr-vs-reynolds_marker.RData"))
load(here(output_data_path, "integr-vs-06_marker.RData"))
load(here(output_data_path, "integr-vs-08_marker.RData"))

```


```{r, include=FALSE,  fig.width = 12, fig.height = 5}
# de_integr_reynolds <- find_de_genes(list("obj_integr" = obj_integr, "obj_reynolds" = obj_reynolds), "predicted.id.Cell_type_nicknames", output_data_path)
# save(de_integr_reynolds, file = here(output_data_path, "integr-vs-reynolds_marker.RData"))
filename <- here(output_data_path, "integr-vs-reynolds_marker.xlsx")
de_integr_reynolds_sign <- map(de_integr_reynolds,~filter(., p_val_adj <0.05))
wb <- createWorkbook()
for(i in seq_along(de_integr_reynolds))
{
  gc()
  message("Creating sheet", i)
  sheet <- createSheet(wb, sheetName = names(de_integr_reynolds)[i])
  message("Adding data frame", i)
  addDataFrame(de_integr_reynolds[[i]], sheet)
}
saveWorkbook(wb, filename)
gc()
# de_integr_06 <- find_de_genes(list("obj_integr" = obj_integr, "obj_06" = obj_06), "predicted.id.Cell_type_nicknames", output_data_path)
# save(de_integr_06, file = here(output_data_path, "integr-vs-06_marker.RData"))
# filename <- here(output_data_path, "integr-vs-06_marker.xlsx")
# de_integr_06_sign <- map(de_integr_06,~filter(., p_val_adj <0.05))
# wb <- createWorkbook()
# for(i in seq_along(de_integr_06))
# {
#   gc()
#   message("Creating sheet", i)
#   sheet <- createSheet(wb, sheetName = names(de_integr_06)[i])
#   message("Adding data frame", i)
#   addDataFrame(de_integr_06[[i]], sheet)
# }
# saveWorkbook(wb, filename)
# gc()
# de_integr_08 <- find_de_genes(list("obj_integr" = obj_integr, "obj_08" = obj_08), "predicted.id.Cell_type_nicknames", output_data_path)
# save(de_integr_08, file = here(output_data_path, "integr-vs-08_marker.RData"))
# filename <- here(output_data_path, "integr-vs-08_marker.xlsx")
# de_integr_08_sign <- map(de_integr_08,~filter(., p_val_adj <0.05))
# wb <- createWorkbook()
# for(i in seq_along(de_integr_08))
# {
#     gc()
#   message("Creating sheet", i)
#   sheet <- createSheet(wb, sheetName = names(de_integr_08)[i])
#   message("Adding data frame", i)
#   addDataFrame(de_integr_08[[i]], sheet)
# }
# saveWorkbook(wb, filename)
# gc()

```
# {.tabset}

## Dataset BF_01+02+03 vs reynolds data 


```{r, fig.height=15, fig.width=20}
obj_reynolds$predicted.id.Cell_type <- obj_reynolds$Cell_type
obj_merged_reynolds <- merge(obj_integr, obj_reynolds)
DefaultAssay(obj_merged_reynolds) <- "RNA"
obj_merged_reynolds <- NormalizeData(obj_merged_reynolds)
obj_merged_reynolds <- FindVariableFeatures(obj_merged_reynolds)
obj_merged_reynolds <- ScaleData(obj_merged_reynolds)


obj_merged_reynolds$predicted.id.Cell_type_nicknames <- make_nicknames(obj_merged_reynolds$predicted.id.Cell_type)

obj_merged_reynolds$temp.ident <- "objreynolds"
obj_merged_reynolds$temp.ident[grep("BF", colnames(obj_merged_reynolds))] <- "objintegr"
DefaultAssay(obj_merged_reynolds) <- "RNA"

de_integr_reynolds <- map(de_integr_reynolds, ~rownames_to_column(., var = "feature")) %>%
  map( ~as_tibble(.))

```

### Top-5 differential expression features across inflammatory celltypes

```{r, fig.height=10, fig.width=15}
dendritic <- grep("DC", unique(obj_integr$predicted.id.Cell_type_nicknames), value = TRUE)
tcells <- c("Tc", "Th", "Treg")
macrophages <- grep("Mac", unique(obj_integr$predicted.id.Cell_type_nicknames), value = TRUE)
langerhans <- grep("LC", unique(obj_integr$predicted.id.Cell_type_nicknames), value = TRUE)
inflammatory <- sort(c(dendritic, tcells, macrophages, langerhans, "Mast cell", "Monocyte", "Plasma cell" ))

top10_marker_inflammatory <- map(de_integr_reynolds,~top_n(., n = 10, wt = avg_log2FC))
top10_marker_inflammatory <- top10_marker_inflammatory[names(top10_marker_inflammatory) %in% inflammatory]
top5_marker_inflammatory_plot_up  <- map(de_integr_reynolds,~top_n(., n = 5, wt = avg_log2FC)) %>%
  map(~pull(., feature)) %>%
  unlist()

top5_marker_inflammatory_plot_down  <- map(de_integr_reynolds,~top_n(., n = -5, wt = avg_log2FC)) %>%
  map(~pull(., feature)) %>%
  unlist()

Idents(obj_merged_reynolds) <- "predicted.id.Cell_type_nicknames"
obj_merged_reynolds_inflammatory <- subset(obj_merged_reynolds, idents = inflammatory)

DotPlot(obj_merged_reynolds_inflammatory, features = unique(top5_marker_inflammatory_plot_up), group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
) +
  ggtitle("up in obj_integr")

DotPlot(obj_merged_reynolds_inflammatory, features = unique(top5_marker_inflammatory_plot_down), group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)+
  ggtitle("down in obj_integr")


```

### Expression of all interleukins 

```{r, fig.height=10, fig.width=20}
interleukins <- intersect(interleukins, rownames(obj_integr))
interleukins <- unique(c(interleukins, rownames(obj_integr)[grep("^IL", rownames(obj_integr))]))

DotPlot(obj_merged_reynolds_inflammatory, features = interleukins, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)
```

### Expression of interestin genes


```{r, fig.height=10, fig.width=15}
genes.bennet <- intersect(c(rownames(obj_integr)[grep("IL17", rownames(obj_integr))], "TNF", "TNFRSF1A", "IL22", rownames(obj_integr)[grep("IL12", rownames(obj_integr))], "IL23R", "IL23A", "NOS1", "NOS2", "IL13", "IL13RA1"), rownames(obj_integr))

DotPlot(obj_merged_reynolds_inflammatory, features = genes.bennet, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)

```


## Dataset BF_01+02+03 vs BF_06 data 

```{r, fig.height=15, fig.width=20}
obj_merged_06 <- merge(obj_integr, obj_06)
DefaultAssay(obj_merged_06) <- "RNA"
obj_merged_06 <- NormalizeData(obj_merged_06)
obj_merged_06 <- FindVariableFeatures(obj_merged_06)
obj_merged_06 <- ScaleData(obj_merged_06)


obj_merged_06$predicted.id.Cell_type_nicknames <- make_nicknames(obj_merged_06$predicted.id.Cell_type)

obj_merged_06$temp.ident <- "objintegr"
obj_merged_06$temp.ident[grep("BF_LE_06" , colnames(obj_merged_06))] <- "obj06"
DefaultAssay(obj_merged_06) <- "RNA"

de_integr_06 <- map(de_integr_06, ~rownames_to_column(., var = "feature")) %>%
  map( ~as_tibble(.))

```

### Top-5 differential expression features across inflammatory celltypes

```{r, fig.height=10, fig.width=15}

top10_marker_inflammatory <- map(de_integr_06,~top_n(., n = 10, wt = avg_log2FC))
top10_marker_inflammatory <- top10_marker_inflammatory[names(top10_marker_inflammatory) %in% inflammatory]
top5_marker_inflammatory_plot_up  <- map(de_integr_06,~top_n(., n = 5, wt = avg_log2FC)) %>%
  map(~pull(., feature)) %>%
  unlist()

top5_marker_inflammatory_plot_down  <- map(de_integr_06,~top_n(., n = -5, wt = avg_log2FC)) %>%
  map(~pull(., feature)) %>%
  unlist()

Idents(obj_merged_06) <- "predicted.id.Cell_type_nicknames"
obj_merged_06_inflammatory <- subset(obj_merged_06, idents = inflammatory)

DotPlot(obj_merged_06_inflammatory, features = unique(top5_marker_inflammatory_plot_up), group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
) +
  ggtitle("up in obj_integr")

DotPlot(obj_merged_06_inflammatory, features = unique(top5_marker_inflammatory_plot_down), group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
) +
  ggtitle("down in obj_integr")


```

### Expression of all interleukins 

```{r, fig.height=10, fig.width=20}
interleukins <- intersect(interleukins, rownames(obj_integr))
interleukins <- unique(c(interleukins, rownames(obj_integr)[grep("^IL", rownames(obj_integr))]))

DotPlot(obj_merged_06_inflammatory, features = interleukins, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)
```

### Expression of interestin genes


```{r, fig.height=10, fig.width=15}
genes.bennet <- intersect(c(rownames(obj_integr)[grep("IL17", rownames(obj_integr))], "TNF", "TNFRSF1A", "IL22", rownames(obj_integr)[grep("IL12", rownames(obj_integr))], "IL23R", "IL23A", "NOS1", "NOS2", "IL13", "IL13RA1"), rownames(obj_integr))

DotPlot(obj_merged_06_inflammatory, features = genes.bennet, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)

```




## Dataset BF_01+02+03 vs BF_08 data 

```{r, fig.height=15, fig.width=20}
obj_merged_08 <- merge(obj_integr, obj_08)
DefaultAssay(obj_merged_08) <- "RNA"
obj_merged_08 <- NormalizeData(obj_merged_08)
obj_merged_08 <- FindVariableFeatures(obj_merged_08)
obj_merged_08 <- ScaleData(obj_merged_08)


obj_merged_08$predicted.id.Cell_type_nicknames <- make_nicknames(obj_merged_08$predicted.id.Cell_type)

obj_merged_08$temp.ident <- "objintegr"
obj_merged_08$temp.ident[grep("BF_LE_08" , colnames(obj_merged_08))] <- "obj08"
DefaultAssay(obj_merged_08) <- "RNA"


de_integr_08 <- map(de_integr_08, ~rownames_to_column(., var = "feature")) %>%
  map( ~as_tibble(.))

```

### Top-5 differential expression features across inflammatory celltypes

```{r, fig.height=10, fig.width=15}

top10_marker_inflammatory <- map(de_integr_08,~top_n(., n = 10, wt = avg_log2FC))
top10_marker_inflammatory <- top10_marker_inflammatory[names(top10_marker_inflammatory) %in% inflammatory]
top5_marker_inflammatory_plot_up  <- map(de_integr_08,~top_n(., n = 5, wt = avg_log2FC)) %>%
  map(~pull(., feature)) %>%
  unlist()

top5_marker_inflammatory_plot_down  <- map(de_integr_08,~top_n(., n = -5, wt = avg_log2FC)) %>%
  map(~pull(., feature)) %>%
  unlist()

Idents(obj_merged_08) <- "predicted.id.Cell_type_nicknames"
obj_merged_08_inflammatory <- subset(obj_merged_08, idents = inflammatory)

DotPlot(obj_merged_08_inflammatory, features = unique(top5_marker_inflammatory_plot_up), group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
) +
  ggtitle("up in obj_integr")

DotPlot(obj_merged_08_inflammatory, features = unique(top5_marker_inflammatory_plot_down), group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
) +
  ggtitle("down in obj_integr")


```

### Expression of all interleukins 

```{r, fig.height=10, fig.width=20}
interleukins <- intersect(interleukins, rownames(obj_integr))
interleukins <- unique(c(interleukins, rownames(obj_integr)[grep("^IL", rownames(obj_integr))]))

DotPlot(obj_merged_08_inflammatory, features = interleukins, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
  )
    
```    
    
    
### Expression of interestin genes


```{r, fig.height=10, fig.width=15}
genes.bennet <- intersect(c(rownames(obj_integr)[grep("IL17", rownames(obj_integr))], "TNF", "TNFRSF1A", "IL22", rownames(obj_integr)[grep("IL12", rownames(obj_integr))], "IL23R", "IL23A", "NOS1", "NOS2", "IL13", "IL13RA1"), rownames(obj_integr))

DotPlot(obj_merged_08_inflammatory, features = genes.bennet, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)

```    
    
    
)
```

### Expression of interestin genes


```{r, fig.height=10, fig.width=15}
genes.bennet <- intersect(c(rownames(obj_integr)[grep("IL17", rownames(obj_integr))], "TNF", "TNFRSF1A", "IL22", rownames(obj_integr)[grep("IL12", rownames(obj_integr))], "IL23R", "IL23A", "NOS1", "NOS2", "IL13", "IL13RA1"), rownames(obj_integr))

DotPlot(obj_merged_08_inflammatory, features = genes.bennet, group.by = "predicted.id.Cell_type_nicknames", split.by = "temp.ident", dot.scale = 6, cols = c("red", "blue")) +
  theme(
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10)
)

```

#